src/main/java/TcpServer.java

```diff
@@ -0,0 +1,128 @@
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Optional;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TcpServer implements Closeable {
+  // TODO: make logger instance-specific, prefixing logs w/ config & object info
+  static final Logger logger = Logger.getLogger(TcpServer.class.getName());
+
+  // configuration used to initialize the serverSocket
+  final ServerConfig config;
+
+  // number of clients currently connected to the server socket
+  private final AtomicInteger connectionCount;
+
+  // executor for handling of new connections
+  private final ExecutorService connectionHandlerExecutor;
+
+  // Thread where new connections are accepted from serverSocket
+  private Optional<Thread> acceptThread;
+
+  // socket which manages client connections
+  // Initialized to empty, and set to a Socket object once started.
+  private Optional<ServerSocket> serverSocket;
+
+  TcpServer(ServerConfig config) {
+    this.config = config;
+    this.serverSocket = Optional.empty();
+    this.acceptThread = Optional.empty();
+    this.connectionCount = new AtomicInteger(0);
+    this.connectionHandlerExecutor = Executors.newFixedThreadPool(this.config.maxConnections);
+  }
+
+  /**
+   * @see connectionCount
+    */
+  int getConnectionCount() {
+    return connectionCount.get();
+  }
+
+  /**
+   * Start listening on the specified port.
+   *
+   * Must not be called more than once.
+   *
+   * @throws IOException If the underlying socket couldn't bind successfully.
+   */
+  void start() throws IOException {
```

Comment:
The comment here has a very interesting note "Must not be called more than once." - this made me wonder which other methods are and aren't OK to call multiple times (`close()`?). 

I don't mind JavaDocs, particularly for large projects, but I also don't tend to push for them. In this case, the `throws IOException` is in the method signature already, "Start listening" is covered by the method name, and "on the specified port" confuses me a bit because the port is specified in another object. If you want to keep this comment, I'd say at least make it explicit where the configuration comes from (the `config` field).

Reply:
Maybe this would all be better served with class-level documents, specifically a "usage" section which explains that `TcpServer` objects aren't meant to be reused.